src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -202,9 +284,173 @@ public Task LoadAsync()
+            {
+                if (!CheckIfHasPendingNomination() || cancellationToken.IsCancellationRequested)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _whenNominatedTask ??= new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                if (!_wasSourceBlockContinuationSet)
+                {
+                    _wasSourceBlockContinuationSet = true;
+
+                    _ = SourceBlock.Completion.ContinueWith(t =>
+                    {
+                        lock (SyncObject)
+                        {
+                            if (_whenNominatedTask is not null)
+                            {
+                                if (t.IsFaulted)
+                                {
+                                    _whenNominatedTask.SetException(t.Exception);
+                                }
+                                else
+                                {
+                                    _whenNominatedTask.TrySetCanceled();
+                                }
+                            }
+                        }
+                    }, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
+                }
+            }
+
+            return _whenNominatedTask.Task.WithCancellation(cancellationToken);
+        }
+
+        // True means the project system plans to call NominateProjectAsync in the future.
+        bool IVsProjectRestoreInfoSource.HasPendingNomination
+        {
+            get => CheckIfHasPendingNomination();
+        }
+
+        private bool CheckIfHasPendingNomination()
+        {
+            lock (SyncObject)
+            {
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+                // Nuget should not wait for projects that failed DTB
+                if (!_enabled || SourceBlock.Completion.IsFaulted || SourceBlock.Completion.IsCompleted)
+                {
+                    return false;
+                }
+
+                // Avoid possible deadlock.
+                // Because RestoreCoreAsync() is called inside a dataflow block it will not be called with new data
+                // until the old task finishes. So, if the project gets nominating restore, it will not get updated data.
+                if (IsNuGetRestoreOnGoing())
+                {
+                    return false;
+                }
+
+                ConfiguredProject? activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+                // After the first nomination, we should check the saved nominated version
+                return IsSavedNominationEmptyOrOlder(activeConfiguredProject);
+            }
+        }
+
+        private bool IsNuGetRestoreOnGoing()
```

Comment:
Please use "Package" instead of "NuGet" - its very deliberate that NuGet has not been used throughout this code.

Reply:
Nit: You've used "RestoreOnGoing" and "RestoreStarted" - just pick one name and use it everywhere.

Reply:
method name changed in this [commit](https://github.com/dotnet/project-system/pull/7199/commits/d5a2a1f9b385e6a96d6d1c5fcbde5b7d0f19ebc5)